### PR TITLE
IGAPP-784: Refactor news fetching

### PR DESCRIPTION
**Pull request by klinzo**
This pull request belongs to an issue on our [bugtracker](https://issues.integreat-app.de/).
You can find it there by looking for an issue with the key which is mentioned in the title of this pull request.
It starts with the keyword **IGAPP**.

### CODE REVIEW

File: native/src/components/HeaderContainer.ts [hunk 1]
```diff
@@ -47,6 +47,18 @@ const mapStateToProps = (state: StateType, ownProps: OwnPropsType): StatePropsTy
           route: routeName as typeof simpleRoutes extends (infer U)[] ? U : never
         })
       : null
+  const newsRouteShareUrl =
```
Comment (klinzo) [inline native/src/components/HeaderContainer.ts:50]: Will refactor this in a separate PR but works fine for now (https://issues.tuerantuer.org/browse/IGAPP-635)

File: native/src/routes/NewsContainer.tsx [hunk 2]
```diff
@@ -1,320 +1,78 @@
...
-            criterion: {
-              forceUpdate: false,
-              shouldRefreshResources: false
-            }
-          }
-        }
-        dispatch(fetchNews)
+  const changeUnavailableLanguage = (newLanguage: string) => {
+    setSelectedNewsId(null)
+    setSelectedLanguage(newLanguage)
```
Comment (klinzo) [inline native/src/routes/NewsContainer.tsx:50]: Will refactor this in a separate PR such that we don't have to keep a local state.

**Comment by klinzo** [comment]
More than 1000 LOC deleted :grin: Though I must be honest that I deleted while merely writing some tests, will do so in a separate PR if you think this PR is fine and good to merge.

Explanation what I did in this PR:
* Moved the entire fetching logic of local and tü news to hooks (previously sagas). Reasons for this:
  * The bug I wanted to solve was not solvable with the previous structure (app crashing when selecting an available language for tünews if using an unavailable language)
  * Very complicated logic with lots of workarounds (header, language change)
  * Unnecessary circuit over sagas, redux store and withPayloadProvider event though data is only needed in news
  * Hooks are nicer and a lot simpler (personal preference) :D

Missing:
* Tests
* Refactor and remove ts-ignores from Header
* Language change in news route is missing now (similar to offers and disclaimer)

File: native/src/hooks/useLoadTuNews.ts [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+      })
+    }
+
+    loadFromEndpoint<TunewsModel[]>(request, updateData, setError, setLoading).catch(e => {
+      setError(e)
+    })
+  }, [language, page, updateData])
+
+  const loadMore = useCallback(() => {
+    // Only load more news if not already loading and it is possible that we have more news (i.e. each page is full)
```
Comment (SteffiStuffel) [inline native/src/hooks/useLoadTuNews.ts:61]: Such comments can easily avoided/replaced, by creating a boolean variable with a speaking name. (Just preference though)

File: native/src/routes/News.tsx [hunk 4]
```diff
@@ -8,82 +8,84 @@ import {
...
-  news: NewsModelsType
   cityModel: CityModel
   language: string
   selectedNewsType: NewsType
-  isFetchingMore: boolean
-  fetchMoreNews: () => void
-  navigateTo: (arg0: RouteInformationType) => void
-  routeKey: string
+  loadMore?: () => void
+  loadingMore?: boolean
```
Comment (SteffiStuffel) [inline native/src/routes/News.tsx:39]: Don't see where this is ever set?

File: native/src/routes/NewsContainer.tsx [hunk 5]
```diff
@@ -1,319 +1,86 @@
...
+
+  useEffect(
+    // Handle back navigation: If we are at a news detail screen navigate back to overview instead of closing the route
+    () =>
+      navigation.addListener('beforeRemove', e => {
+        if (selectedNewsId) {
+          e.preventDefault()
+          setSelectedNewsId(null)
+        }
+      }),
```
Comment (SteffiStuffel) [inline native/src/routes/NewsContainer.tsx:37]: Should the Listener be removed?

**Review by SteffiStuffel — APPROVED** [review]
Really nice refactoring! Like it. :)

Tested on Android emulator. Works fine, just some design-issues
igapp784

File: native/src/routes/News.tsx (hunk 4 above)
Comment (klinzo) [reply native/src/routes/News.tsx:39]: Set by passing the whole response of `useLoadTuNews` to `News`

File: native/src/routes/NewsContainer.tsx (hunk 5 above)
Comment (klinzo) [reply native/src/routes/NewsContainer.tsx:37]: What do you mean by that? Like completely or just returning the remove function to make sure everything is cleaned up correctly?
If the latter, it actually is already as `navigation.addListener` returns the remove function.

Comment (SteffiStuffel) [reply native/src/routes/NewsContainer.tsx:37]: yeah i meant the cleanup. overlooked it.

File: native/src/components/HeaderContainer.ts [hunk 6]
```diff
@@ -47,6 +47,18 @@ const mapStateToProps = (state: StateType, ownProps: OwnPropsType): StatePropsTy
...
+    ownProps.scene.route.name === NEWS_ROUTE
+      ? urlFromRouteInformation({
+          cityCode,
+          languageCode,
+          route: NEWS_ROUTE,
+          // @ts-ignore
+          newsType: ownProps.scene.route.params.newsType,
+          // @ts-ignore
+          newsId: ownProps.scene.route.params.newsId
+        })
```
Comment (Taggotty) [inline native/src/components/HeaderContainer.ts:60]: sharing did not work for me for tunews content. For me it created this link: https://integreat.app/testumgebung/de/news/local

**Review by Taggotty — COMMENTED** [review]
Nice work. 🥳 
Tested on Android emulator. Same design issue as @SteffiStuffel mentioned.

**Comment by klinzo** [comment]
@SteffiStuffel @Taggotty The design issues are only for that specific tünews and did already exist before my PR. Please create an issue :)

File: native/src/components/HeaderContainer.ts (hunk 1 above)
Comment (klinzo) [reply native/src/components/HeaderContainer.ts:50]: See https://github.com/Integreat/integreat-app/pull/524

File: native/src/components/HeaderContainer.ts (hunk 6 above)
Comment (klinzo) [reply native/src/components/HeaderContainer.ts:60]: Done in #524

**Review by Taggotty — APPROVED** [review]

**Comment by klinzo** [comment]
Created an issue for the design problems: https://issues.tuerantuer.org/browse/IGAPP-863

**Pull request merged**